src/v/cluster/partition_balancer_planner.cc

```diff
@@ -865,6 +867,42 @@ void partition_balancer_planner::get_rack_constraint_repair_actions(
     }
 }
 
+/**
+ * This is the place where we decide about the order in which partitions will be
+ * moved in the case when node disk is being full.
+ */
+size_t partition_balancer_planner::calculate_full_disk_partition_move_priority(
+  model::node_id node_id,
+  const reassignable_partition& p,
+  const request_context& ctx) {
+    static constexpr size_t min_priority = 0;
+    static constexpr size_t max_priority = 1000000;
+
+    if (p.size_bytes() < ctx.config().min_partition_size_threshold) {
+        return min_priority;
+    }
+    /**
+     * Use lowest priority + 1 for redpanda internal topics, this way large
+     * internal topics will be moved before smallest user partitions
+     */
+    if (
+      p.ntp().ns == model::kafka_internal_namespace
+      || p.ntp().tp.topic == model::kafka_consumer_offsets_topic) {
+        return min_priority + 1;
+    }
+
+    auto it = ctx.node_disk_reports.find(node_id);
+    if (it == ctx.node_disk_reports.end()) {
+        return min_priority;
+    }
+
+    // normalize to range between [0,1000000], where max value would represent a
+    // partition that is of the full disk capacity size. We subtract it from the
+    // max priority to prioritize smallest partitions.
+    return max_priority
+           - (max_priority * p.size_bytes()) / ctx.node_disk_reports.capacity();
```

Comment:
Hmm, this is hash table capacity, right? Probably not what you meant :)

Reply:
great catch